core/src/main/scala/kafka/tools/EndToEndLatency.scala

```diff
@@ -58,35 +69,57 @@ object EndToEndLatency {
     producerProps.put(ProducerConfig.ACKS_CONFIG, producerAcks.toString)
     producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
     producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
-    val producer = new KafkaProducer[Array[Byte],Array[Byte]](producerProps)
+    val producer = new KafkaProducer[Array[Byte], Array[Byte]](producerProps)
+
+    def finalise() {
+      consumer.commit(CommitType.SYNC)
+      producer.close()
+      consumer.close()
+    }
 
-    // make sure the consumer fetcher has started before sending data since otherwise
-    // the consumption from the tail will skip the first message and hence be blocked
-    Thread.sleep(5000)
+    //Ensure we are at latest offset
+    var recordIter = consumer.poll(0).iterator
+    consumer.seekToEnd()
 
-    val message = "hello there beautiful".getBytes
     var totalTime = 0.0
     val latencies = new Array[Long](numMessages)
+
     for (i <- 0 until numMessages) {
+      val message = arrayOfLen(messageLen)
       val begin = System.nanoTime
-      producer.send(new ProducerRecord[Array[Byte],Array[Byte]](topic, message))
-      val received = iter.next
+
+      producer.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, message))
+      recordIter = consumer.poll(Long.MaxValue).iterator
```

Comment:
We shouldn't get more than one message at a time, assuming other processes aren't writing to the topic during this test. This seems reasonable to me but I'm happy to take advice on this. I've forced the test to fail should this happen.